core/node/node_framework/src/service/tests.rs

```diff
@@ -0,0 +1,232 @@
+#[test]
+fn test_run_with_failed_tasks() {
+    let mut zk_stack_service: ZkStackService = ZkStackService::new().unwrap();
+    zk_stack_service.add_layer(TaskErrorLayer);
+    let result = zk_stack_service.run();
+    assert_eq!(
+        result.unwrap_err().to_string(),
+        "Task error_task failed".to_string(),
+        "Incorrect result for creating a service with task that fails"
+    );
+}
+
+// Layer that adds a task that updates resource and returns
+#[derive(Debug)]
+struct SuccessfulTaskLayer(Arc<Mutex<bool>>);
+
+#[async_trait::async_trait]
+impl WiringLayer for SuccessfulTaskLayer {
+    fn layer_name(&self) -> &'static str {
+        "successful_task_layer"
+    }
+
+    async fn wire(self: Box<Self>, mut node: ServiceContext<'_>) -> Result<(), WiringError> {
+        node.add_task(Box::new(SuccessfulTask(self.0.clone())));
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct SuccessfulTask(Arc<Mutex<bool>>);
+
+#[async_trait::async_trait]
+impl Task for SuccessfulTask {
+    fn name(&self) -> &'static str {
+        "successful_task"
+    }
+    async fn run(self: Box<Self>, _stop_receiver: StopReceiver) -> anyhow::Result<()> {
+        let mut guard = self.0.lock().unwrap();
+        *guard = true;
+        Ok(())
+    }
+}
+
+// Layer for task that updates resource and returns only after stop signal received
+#[derive(Debug)]
+struct RemainingTaskLayer(Arc<Mutex<bool>>);
+
+#[async_trait::async_trait]
+impl WiringLayer for RemainingTaskLayer {
+    fn layer_name(&self) -> &'static str {
+        "remaining_task_layer"
+    }
+
+    async fn wire(self: Box<Self>, mut node: ServiceContext<'_>) -> Result<(), WiringError> {
+        node.add_task(Box::new(RemainingTask(self.0.clone())));
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct RemainingTask(Arc<Mutex<bool>>);
+
+#[async_trait::async_trait]
+impl Task for RemainingTask {
+    fn name(&self) -> &'static str {
+        "remaining_task"
+    }
+    async fn run(self: Box<Self>, mut stop_receiver: StopReceiver) -> anyhow::Result<()> {
+        stop_receiver.0.changed().await?;
+        let mut guard = self.0.lock().unwrap();
+        *guard = false;
+        Ok(())
+    }
+}
+
+// Check ZkStack Service's `run()` method expected behavior.
+// There are 2 tests, one for successful task run and another for remaining task run.
+// Second test case relies on the first one.
+#[test]
+fn test_task_run() {
+    // case 1: run the task and check it updates resource successfully before stop signal received.
+    let some_bool_resource = Arc::new(Mutex::new(false));
+    let mut zk_stack_service = ZkStackService::new().unwrap();
+    zk_stack_service.add_layer(SuccessfulTaskLayer(some_bool_resource.clone()));
+
+    assert!(
+        zk_stack_service.run().is_ok(),
+        "ZkStackService run failed, but it shouldn't"
+    );
+    let actual_resource_value = *some_bool_resource.lock().unwrap();
```

Comment:
I think we can find a name for this variable that would represent the logical meaning.
In this case, for example, how about `task_was_run` (because the flag is set when `run()` method is invoked)?